cmd/kes/identity.go

```diff
@@ -120,142 +112,162 @@ func newIdentityCmd(args []string) {
+			func(cert *x509.Certificate) { cert.IPAddresses = IPs },
 		}
-		if _, err = os.Stat(certPath); err == nil {
-			cli.Fatal("certificate already exists. Use --force to overwrite it")
+		if cmd.NArg() == 1 {
+			name := cmd.Arg(0)
+			options = append(options, func(cert *x509.Certificate) { cert.Subject.CommonName = name })
 		}
-	}
-
-	var keyPem []byte
-	if encrypt {
-		fmt.Fprint(os.Stderr, "Enter password for private key:")
-		p, err := term.ReadPassword(int(os.Stderr.Fd()))
-		if err != nil {
-			cli.Fatal(err)
+		if expiry > 0 {
+			options = append(options, func(cert *x509.Certificate) {
+				now := time.Now()
+				cert.NotBefore, cert.NotAfter = now, now.Add(expiry)
+			})
 		}
-		fmt.Fprintln(os.Stderr)
-		fmt.Fprint(os.Stderr, "Confirm password for private key:")
-		confirm, err := term.ReadPassword(int(os.Stderr.Fd()))
+		cert, err := kes.GenerateCertificate(key, options...)
 		if err != nil {
-			cli.Fatal(err)
-		}
-		fmt.Fprintln(os.Stderr)
-		if !bytes.Equal(p, confirm) {
-			cli.Fatal("passwords don't match")
+			cli.Fatalf("failed to generate certificate: %v", err)
 		}
 
-		block, err := x509.EncryptPEMBlock(rand.Reader, "PRIVATE KEY", privBytes, p, x509.PEMCipherAES256)
+		certBytes := cert.Certificate[0]
+		privBytes, err := x509.MarshalPKCS8PrivateKey(key.Private())
 		if err != nil {
-			cli.Fatalf("failed to encrypt private key: %v", err)
+			cli.Fatalf("failed to create private key: %v", err)
 		}
-		keyPem = pem.EncodeToMemory(block)
-	} else {
-		keyPem = pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: privBytes})
-	}
-	certPem := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: certBytes})
 
-	if err = os.WriteFile(keyPath, keyPem, 0o600); err != nil {
-		cli.Fatalf("failed to create private key: %v", err)
-	}
-	if err = os.WriteFile(certPath, certPem, 0o644); err != nil {
-		os.Remove(keyPath)
-		cli.Fatalf("failed to create certificate: %v", err)
-	}
+		if !forceFlag {
+			if _, err = os.Stat(keyPath); err == nil {
+				cli.Fatal("private key already exists. Use --force to overwrite it")
+			}
+			if _, err = os.Stat(certPath); err == nil {
+				cli.Fatal("certificate already exists. Use --force to overwrite it")
+			}
+		}
 
-	if isTerm(os.Stdout) {
-		fmt.Printf("\n  Private key:  %s\n", keyPath)
-		fmt.Printf("  Certificate:  %s\n", certPath)
+		var keyPem []byte
+		if encrypt {
+			fmt.Fprint(os.Stderr, "Enter password for private key:")
+			p, err := term.ReadPassword(int(os.Stderr.Fd()))
+			if err != nil {
+				cli.Fatal(err)
+			}
+			fmt.Fprintln(os.Stderr)
+			fmt.Fprint(os.Stderr, "Confirm password for private key:")
+			confirm, err := term.ReadPassword(int(os.Stderr.Fd()))
+			if err != nil {
+				cli.Fatal(err)
+			}
+			fmt.Fprintln(os.Stderr)
+			if !bytes.Equal(p, confirm) {
+				cli.Fatal("passwords don't match")
+			}
+
+			block, err := x509.EncryptPEMBlock(rand.Reader, "PRIVATE KEY", privBytes, p, x509.PEMCipherAES256)
```

Comment:
We don't care about the deprecation notice?

```
// Deprecated: Legacy PEM encryption as specified in RFC 1423 is insecure by
// design. Since it does not authenticate the ciphertext, it is vulnerable to
// padding oracle attacks that can let an attacker recover the plaintext.
```


(though I guess this is also legacy for us)

Reply:
Yes, we just support PEM-encrypted private keys for compatibility with legacy systems requiring password-protected private keys.